blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/ShardStateManager.java

```diff
@@ -317,34 +357,51 @@ protected UpdateStamp getAndSetState(int slot, UpdateStamp.State state) {
                 if (timeElapsed <= maxAgeMillis) {
                     continue;
                 }
-
                 if (update.getLastRollupTimestamp() > 0) {
                     final long timeElapsedSinceLastRollup = now - update.getLastRollupTimestamp();
 
-                    //Handling re-rolls: Re-rolls would use different delay
-                    //Since we only allow delayed metrics upto 3 days(BEFORE_CURRENT_COLLECTIONTIME_MS), a slot can be
-                    //identified as being re-rolled, if the last rollup is within those last 3 days.
+                    //Handling re-rolls: Since we only allow delayed metrics upto 3 days(BEFORE_CURRENT_COLLECTIONTIME_MS),
+                    //a slot can be identified as being re-rolled, if the last rollup is within those last 3 days.
                     if (timeElapsedSinceLastRollup < REROLL_TIME_SPAN_ASSUMED_VALUE) {
 
-                        if (timeElapsed <= delayedMetricsMaxAgeMillis) {
+                        //short delay
+                        SlotKey slotKey = SlotKey.of(granularity, entry.getKey(), shard);
+                        if (timeElapsed <= rollupDelayForMetricsWithShortDelay) {
 
-                            SlotKey slotKey = SlotKey.of(granularity, entry.getKey(), shard);
-                            log.debug(String.format("Delaying re-roll of slotKey [%s] as [%d] millis haven't elapsed " +
-                                    "since collection time:[%d] now: [%d], last rollup time: [%d]", slotKey,
-                                    delayedMetricsMaxAgeMillis, update.getTimestamp(), now, update.getLastRollupTimestamp()));
+                            reRollForShortDelayMetricsMeters.get(granularity).mark();
+                            log.debug(String.format("Short delay: Delaying re-roll of slotKey [%s] as [%d] millis " +
+                                    "haven't elapsed since collection time:[%d] now: [%d] time elapsed: [%d] last " +
+                                    "rollup time: [%d]", slotKey, rollupDelayForMetricsWithShortDelay,
+                                    update.getTimestamp(), now, timeElapsed, update.getLastRollupTimestamp()));
                             continue;
-                        } else {
+                        }
+
+                        if (update.getLastIngestTimestamp() > 0 ) {
+                            long delayOfLastIngestedMetric = update.getLastIngestTimestamp() - update.getTimestamp();
+                            final long timeElapsedSinceLastIngest = now - update.getLastIngestTimestamp();
 
-                            granToReRollMeters.get(granularity).mark();
-                            if (nowMillis - update.getTimestamp() >= millisInADay) {
-                                granToDelayedMetricsMeter.get(granularity).mark();
+                            //long delay
+                            if (delayOfLastIngestedMetric > rollupDelayForMetricsWithShortDelay &&
```

Comment:
So the first time we cross the short delay(2hr) barrier, we are supposed to rollup. The problem was how do we know its the first time? The condition is, when we receive a delayed metric, we check if it was ingested within the short delay. If it is, its probably the first time.